test/utils/ktesting/contexthelper.go

```diff
@@ -0,0 +1,91 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ktesting
+
+import (
+	"context"
+	"fmt"
+	"time"
+)
+
+// cleanupErr creates a cause when canceling a context because the test has completed.
+// It is a context.Canceled error.
+func cleanupErr(testName string) error {
+	return canceledError(fmt.Sprintf("test %s is cleaning up", testName))
+}
+
+type canceledError string
+
+func (c canceledError) Error() string { return string(c) }
+
+func (c canceledError) Is(target error) bool {
+	return target == context.Canceled
+}
+
+// withTimeout corresponds to [context.WithTimeout]. In contrast to
+// [context.WithTimeout], it automatically cancels during test cleanup, provides
+// the given cause when the deadline is reached, and its cancel function
+// requires a cause.
+func withTimeout(ctx context.Context, tb TB, timeout time.Duration, timeoutCause string) (context.Context, func(cause string)) {
+	tb.Helper()
+
+	now := time.Now()
+
+	cancelCtx, cancel := context.WithCancelCause(ctx)
+	after := time.NewTimer(timeout)
+	stopCtx, stop := context.WithCancel(ctx) // Only used internally, doesn't need a cause.
```

Comment:
Without `stopCtx`, the anonymous function below would block only on `after.C`. It could also block on the parent `ctx.Done()`, but both of these have the same problem: they don't trigger when the test is complete. That would leave the goroutine running after the test is done.

We need the anonymous function for the situation that the timeout triggers while the test still runs.